### PR TITLE
Unify transfer / timer / replication task notification

### PR DESCRIPTION
**Pull request by wxing1292**
* Unify transfer / replication / timer task notification, now all notifications are handled in workflow execution context
* Remove unnecessary GetTransferTasks & GetTimerTasks in mutable state
* Remove unnecessary shardContextWrapper

### CODE REVIEW

File: common/persistence/persistence-tests/persistenceTestBase.go [hunk 1]
```diff
@@ -44,7 +44,7 @@ import (
 type (
 	// TransferTaskIDGenerator generates IDs for transfer tasks written by helper methods
 	TransferTaskIDGenerator interface {
-		GetNextTransferTaskID() (int64, error)
+		GetTransferTaskID() (int64, error)
```
Comment (vancexu) [inline common/persistence/persistence-tests/persistenceTestBase.go]: This feel less clear than previous name

File: service/history/historyEngine2_test.go [hunk 2]
```diff
@@ -142,25 +143,39 @@ func (s *engine2Suite) SetupTest() {
...
+		historyMgr:           s.mockHistoryMgr,
+		historyV2Mgr:         s.mockHistoryV2Mgr,
+		historyCache:         historyCache,
+		logger:               s.logger,
+		metricsClient:        metrics.NewClient(tally.NoopScope, metrics.History),
+		tokenSerializer:      common.NewJSONTaskTokenSerializer(),
+		config:               s.config,
+		archivalClient:       s.mockArchivalClient,
+		timeSource:           mockShard.timeSource,
+		historyEventNotifier: newHistoryEventNotifier(clock.NewRealTimeSource(), metrics.NewClient(tally.NoopScope, metrics.History), func(string) int { return 0 }),
```
Comment (vancexu) [inline service/history/historyEngine2_test.go]: Would be better to reuse metricsClient, and declare the last arg outside for better understanding.

File: service/history/historyEngine3_eventsv2_test.go [hunk 3]
```diff
@@ -136,25 +138,39 @@ func (s *engine3Suite) SetupTest() {
...
+		historyMgr:           s.mockHistoryMgr,
+		historyV2Mgr:         s.mockHistoryV2Mgr,
+		historyCache:         historyCache,
+		logger:               s.logger,
+		metricsClient:        metrics.NewClient(tally.NoopScope, metrics.History),
+		tokenSerializer:      common.NewJSONTaskTokenSerializer(),
+		config:               s.config,
+		archivalClient:       s.mockArchivalClient,
+		timeSource:           mockShard.timeSource,
+		historyEventNotifier: newHistoryEventNotifier(clock.NewRealTimeSource(), metrics.NewClient(tally.NoopScope, metrics.History), func(string) int { return 0 }),
```
Comment (vancexu) [inline service/history/historyEngine3_eventsv2_test.go]: ditto

File: service/history/decisionHandler.go [hunk 4]
```diff
@@ -573,16 +570,11 @@ Update_History_Loop:
 				if err := context.updateWorkflowExecutionAsActive(handler.shard.GetTimeSource().Now()); err != nil {
 					return nil, err
 				}
-				handler.timerProcessor.NotifyNewTimers(handler.currentClusterName, handler.shard.GetCurrentTime(handler.currentClusterName), timerTasks)
 			}
+
```
Comment (yux0) [inline service/history/decisionHandler.go]: remove empty line

File: service/history/historyEngine.go [hunk 5]
```diff
@@ -82,16 +83,6 @@ type (
 		archivalClient       warchiver.Client
 		resetor              workflowResetor
 	}
-
-	// shardContextWrapper wraps ShardContext to notify transferQueueProcessor on new tasks.
-	// TODO: use to notify timerQueueProcessor as well.
-	shardContextWrapper struct {
```
Comment (yux0) [inline service/history/historyEngine.go]: :+1:

**Pull request merged**